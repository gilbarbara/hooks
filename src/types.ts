import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
