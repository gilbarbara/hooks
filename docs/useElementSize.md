# useElementSize
Get element dimensions using the [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) API.   
Returns an `ElementSize` object.

## Usage

```tsx
import React from 'react';
import { useElementSize } from '@gilbarbara/hooks';

function Component() {
  const ref = React.useRef<HTMLDivElement>(null);
  const dimensions = useElementSize(ref);

  return (
    <div ref={ref}>
      {JSON.stringify(dimensions, undefined, 2)}
    </div>
  );
}
```

## Reference

```typescript
interface ElementSize {
  height: number;
  innerHeight: number;
  innerWidth: number;
  width: number;
}

useElementSize<T extends Element>(
  target: RefObject<T> | T | null | string,
  debounce = 0,
): ElementSize;
```

> This hook uses a [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver), so if you want to support
legacy browsers, install the [resize-observer-polyfill](https://www.npmjs.com/package/resize-observer-polyfill) in your app.
