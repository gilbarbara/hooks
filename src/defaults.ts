import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
