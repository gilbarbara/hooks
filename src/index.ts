export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
