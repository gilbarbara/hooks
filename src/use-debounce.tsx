import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
