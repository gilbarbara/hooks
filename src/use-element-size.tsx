import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
