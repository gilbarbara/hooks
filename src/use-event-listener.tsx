import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
