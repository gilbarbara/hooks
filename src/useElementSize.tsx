import { useState } from 'react';

import { Target } from './types';
import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
import { useResizeObserver } from './useResizeObserver';
import { getElement } from './utils';

export interface ElementSize {
  height: number;
  innerHeight: number;
  innerWidth: number;
  width: number;
}

function parseFloatValue(value: string) {
  const parsed = parseFloat(value);

  return Number.isNaN(parsed) ? 0 : parsed;
}

function getElementSize(element: Element | null): ElementSize {
  if (!element) {
    return {
      height: 0,
      innerHeight: 0,
      innerWidth: 0,
      width: 0,
    };
  }

  const {
    borderBottom,
    borderLeft,
    borderRight,
    borderTop,
    height,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
    width,
  } = getComputedStyle(element);

  return {
    height: parseFloatValue(height),
    innerHeight:
      parseFloatValue(height) -
      parseFloatValue(paddingTop) -
      parseFloatValue(paddingBottom) -
      parseFloatValue(borderTop) -
      parseFloatValue(borderBottom),
    innerWidth:
      parseFloatValue(width) -
      parseFloatValue(paddingLeft) -
      parseFloatValue(paddingRight) -
      parseFloatValue(borderLeft) -
      parseFloatValue(borderRight),
    width: parseFloatValue(width),
  };
}

export function useElementSize<T extends Element>(target: Target<T>, debounce = 0): ElementSize {
  const [element, setElement] = useState(getElement(target));
  const [dimensions, setDimensions] = useState<ElementSize>(getElementSize(element));

  const entry = useResizeObserver(element, debounce);

  useIsomorphicLayoutEffect(() => {
    const nextElement = getElement(target);

    setElement(nextElement);
    setDimensions(getElementSize(nextElement));
  }, [target]);

  useIsomorphicLayoutEffect(() => {
    if (!entry) {
      return;
    }

    const { height, width } = entry.contentRect;
    const { blockSize, inlineSize } = entry.borderBoxSize[0];

    setDimensions({ height: blockSize, innerHeight: height, innerWidth: width, width: inlineSize });
  }, [entry]);

  return dimensions;
}
