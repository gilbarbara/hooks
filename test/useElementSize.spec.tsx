import { MutableRefObject } from 'react';
import { act, renderHook } from '@testing-library/react';
import { mockResizeObserver } from 'jsdom-testing-mocks';

import resizeObserverResponse from './__fixtures__/resizeObserverResponse.json';

import { ElementSize, useElementSize } from '../src/useElementSize';

const resizeObserver = mockResizeObserver();

function createElement(id = 'root') {
  const rootElement = document.createElement('div');

  rootElement.id = id;
  rootElement.innerHTML = 'Hello World';

  return rootElement;
}

const computeStyle = {
  borderBottom: '0px',
  borderLeft: '0px',
  borderRight: '0px',
  borderTop: '0px',
  height: '100px',
  paddingBottom: '0px',
  paddingLeft: '0px',
  paddingRight: '0px',
  paddingTop: '0px',
  width: '200px',
};

const mockGetComputedStyle = jest.fn().mockImplementation(() => ({}));

describe('useElementSize', () => {
  const { getComputedStyle } = window;
  const ref = { current: null } as MutableRefObject<Element | null>;
  const rootElement = createElement();

  beforeAll(() => {
    window.getComputedStyle = mockGetComputedStyle;

    document.body.appendChild(rootElement);
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    window.getComputedStyle = getComputedStyle;

    document.body.removeChild(rootElement);
  });

  it.each([{ target: 'element' }, { target: '#root' }, { target: 'ref' }])(
    'should return the dimensions with a "$target" target',
    ({ target }) => {
      let result: MutableRefObject<ElementSize> | null = null;
      let rerender: () => void;

      if (target === 'ref') {
        ref.current = rootElement;
        ({ rerender, result } = renderHook(() => useElementSize(ref)));
      } else if (target.startsWith('#')) {
        ({ rerender, result } = renderHook(() => useElementSize(target)));
      } else {
        ({ rerender, result } = renderHook(() => useElementSize(rootElement)));
      }

      expect(result?.current).toMatchSnapshot('initial');

      resizeObserver.mockElementSize(rootElement, resizeObserverResponse);

      act(() => {
        resizeObserver.resize();
      });

      rerender();
      expect(result.current).toMatchSnapshot('after resize');

      const { current } = result;

      rerender();
      expect(result.current).toBe(current);
    },
  );

  it('should re-initialize if target changes', async () => {
    const { rerender, result } = renderHook((target: string = '#app') => useElementSize(target));

    expect(result.current).toMatchSnapshot('initial');
    mockGetComputedStyle.mockImplementation(() => computeStyle);
    resizeObserver.mockElementSize(rootElement, resizeObserverResponse);

    rerender('#root');

    act(() => {
      resizeObserver.resize();
    });

    expect(result.current).toMatchSnapshot('after render');

    const { current } = result;

    rerender('#root');
    expect(result.current).toBe(current);
  });

  it('should return the default dimensions with an invalid string selector', () => {
    const { rerender, result } = renderHook(() => useElementSize('#app'));

    expect(result.current).toMatchSnapshot('initial');

    const { current } = result;

    rerender('#app');
    expect(result.current).toMatchSnapshot('after render');
    expect(result.current).toBe(current);
  });
});
