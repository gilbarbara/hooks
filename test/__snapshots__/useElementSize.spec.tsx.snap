// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useElementSize should re-initialize if target changes: after render 1`] = `
{
  "height": 200,
  "innerHeight": 200,
  "innerWidth": 200,
  "width": 200,
}
`;

exports[`useElementSize should re-initialize if target changes: initial 1`] = `
{
  "height": 0,
  "innerHeight": 0,
  "innerWidth": 0,
  "width": 0,
}
`;

exports[`useElementSize should return the default dimensions with an invalid string selector: after render 1`] = `
{
  "height": 0,
  "innerHeight": 0,
  "innerWidth": 0,
  "width": 0,
}
`;

exports[`useElementSize should return the default dimensions with an invalid string selector: initial 1`] = `
{
  "height": 0,
  "innerHeight": 0,
  "innerWidth": 0,
  "width": 0,
}
`;

exports[`useElementSize should return the dimensions with a "#root" target: after resize 1`] = `
{
  "height": 200,
  "innerHeight": 200,
  "innerWidth": 200,
  "width": 200,
}
`;

exports[`useElementSize should return the dimensions with a "#root" target: initial 1`] = `
{
  "height": 0,
  "innerHeight": 0,
  "innerWidth": 0,
  "width": 0,
}
`;

exports[`useElementSize should return the dimensions with a "element" target: after resize 1`] = `
{
  "height": 200,
  "innerHeight": 200,
  "innerWidth": 200,
  "width": 200,
}
`;

exports[`useElementSize should return the dimensions with a "element" target: initial 1`] = `
{
  "height": 0,
  "innerHeight": 0,
  "innerWidth": 0,
  "width": 0,
}
`;

exports[`useElementSize should return the dimensions with a "ref" target: after resize 1`] = `
{
  "height": 200,
  "innerHeight": 200,
  "innerWidth": 200,
  "width": 200,
}
`;

exports[`useElementSize should return the dimensions with a "ref" target: initial 1`] = `
{
  "height": 0,
  "innerHeight": 0,
  "innerWidth": 0,
  "width": 0,
}
`;
